sequence_processing_pipeline/ConvertJob.py

```diff
@@ -210,3 +222,152 @@ def parse_job_script(job_script_path):
+        self.controls_info = controls_info
+
+    def copy_sequences(self, sample_name, source_project, dest_project):
+        """
+        Copies all fastq files related to a sample into another project.
+        :param sample_name: A sample_name value if copy_all_replicates is
+            False; an orig_name value otherwise.
+        :param source_project: The source project name including qiita_id.
+        :param dest_project: The destination project name including qiita_id.
+        :return: None
+        """
+        if self.info is None:
+            raise ValueError("This method cannot be called until processing "
+                             "has completed.")
+
+        project_names = list(self.info.keys())
+
+        # confirm source and destination projects are both valid.
+        for proj in [source_project, dest_project]:
+            if proj not in project_names:
+                raise ValueError(f"'{proj}' is not defined in sample sheet")
+
+        if source_project == dest_project:
+            raise ValueError(f"source '{source_project}' and destination "
+                             f"'{dest_project}' projects are the same")
+
+        # note that the user can supply a sample-name that didn't make it
+        # through the conversion step and may have no files matched to it.
+        # this is considered okay and a possible outcome of conversion. In
+        # this case zero files are copied and no information is passed back
+        # to the user.
+
+        # projects that contain replicates must also be considered. In this
+        # situation, the sample_name needs to be compared against all values
+        # in the orig_name column and the individual sample_names (w/well-ids)
+        # must be discovered; those individual sample_names can then be
+        # processed.
+
+        samples = self.info[source_project][SAMPLES_DETAILS_KEY]
+
+        # Either we will need to copy all replicates, in which case we need to
+        # match our input to orig_name values, or we want to copy
+        # a single sample, in which case the input needs to be matched to
+        # sample_name values. The results of our match will be
+        # stored in results and if we are not copying all replicates, there
+        # should only be one sample matched in results.
+        results = []
+
+        if self.contains_replicates is True:
+            for _, sample in samples.items():
+                # assume orig_name is present if copy_all_replicates is True.
+                if sample_name == sample[ORIG_NAME_KEY]:
+                    results.append(sample)
+        else:
+            # sample_name is a value from the sample_name column. it may or
+            # may not have a well-id appended and this sample-sheet may or
+            # may not contain replicates, but in either case a single sample
+            # either exists or it doesn't.
+            if sample_name in self.info[source_project][SAMPLES_DETAILS_KEY]:
```

Comment:
The branch on line 347 indicates that zero matches is erroneous, as `len(results) == 0` raises. Is it erroneous if any sample doesn't match? Isn't it bad if we cannot relate a sample back to its project information? Or maybe I don't udnetsand something here

Reply:
@wasade , could you clarify this point for me a bit?  I agree that if no sample matches the input `sample_name` (if `len(results)==0`), then the situation is erroneous.  I believe we do error in that situation; the error message text is slightly different depending on whether there are replicates or not, but either way if no match (result) is found, there is a ValueError raised.  Can you say more about what situation is being overlooked?

Reply:
Oh I see, sorry, in this case `sample_name` is singular so at most `len(results)` will be 1.